### PR TITLE
Fixed XPack Set User Passwords failure (#266)

### PR DESCRIPTION
**Pull request by ivan-sam**
Allow comma separated values and special values for network host (#266)
https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-network.html#network-interface-values

### CODE REVIEW

File: src/InstallerHosts/Elastic.InstallerHosts/Elasticsearch/Tasks/Install/SetupXPackPasswordsTask.cs [hunk 1]
```diff
@@ -111,5 +119,63 @@ private void SetPassword(HttpClient client, string elasticUserPassword, string u
...
+					if (host == SiteAddressV6Placeholder || host == GlobalAddressV6Placeholder)
+						ipAddresses = ipAddresses.Where(a => a.AddressFamily == AddressFamily.InterNetworkV6);
+					var ipAddress = ipAddresses.First();
+					host = ipAddress.AddressFamily == AddressFamily.InterNetworkV6 ? $"[{ipAddress}]" : ipAddress.ToString();
+					break;
+			}
+
+			// do not remove trailing slash. Base address *must* have it
+			var baseAddress = $"http://{host}:{port}/";
+			if (!Uri.IsWellFormedUriString(baseAddress, UriKind.Absolute))
```
Comment (Mpdreamz) [inline src/InstallerHosts/Elastic.InstallerHosts/Elasticsearch/Tasks/Install/SetupXPackPasswordsTask.cs]: 👍 love this fallback

File: src/InstallerHosts/Elastic.InstallerHosts/Elasticsearch/Tasks/Install/SetupXPackPasswordsTask.cs [hunk 2]
```diff
@@ -111,5 +119,63 @@ private void SetPassword(HttpClient client, string elasticUserPassword, string u
...
+			var host = networkHost;
+			if (networkHost.IndexOf(',') >= 0) // Uri host must not contain comma, but elasticsearch.yml supports it
+			{
+				var hosts = networkHost.Split(new[] {','}, StringSplitOptions.RemoveEmptyEntries).Select(h => h.Trim());
+				host = hosts.FirstOrDefault(h => h == "localhost" || h == "127.0.0.1"
+					|| h == LocalhostPlaceholder || h == LocalhostV4Placeholder || h == LocalhostV6Placeholder) // prefer local interface
+					?? hosts.First();
+			}
+
+			switch (host)
```
Comment (Mpdreamz) [inline src/InstallerHosts/Elastic.InstallerHosts/Elasticsearch/Tasks/Install/SetupXPackPasswordsTask.cs]: This is great, it attempts to respect the special placeholder syntax from elasticsearch when bootstrapping http calls.

File: src/InstallerHosts/Elastic.InstallerHosts/Elasticsearch/Tasks/Install/SetupXPackPasswordsTask.cs [hunk 3]
```diff
@@ -111,5 +119,63 @@ private void SetPassword(HttpClient client, string elasticUserPassword, string u
 			}
 			this.Session.SendProgress(200, $"Changed password for user '{user}'");
 		}
+
+		// Might be moved and refactored when is needed for other tasks
```
Comment (Mpdreamz) [inline src/InstallerHosts/Elastic.InstallerHosts/Elasticsearch/Tasks/Install/SetupXPackPasswordsTask.cs]:
```suggestion
		// TODO: Might be moved and refactored when is needed for other tasks
```

Use todo syntax so it shows up in Todo explorer.

File: src/InstallerHosts/Elastic.InstallerHosts/Elasticsearch/Tasks/Install/SetupXPackPasswordsTask.cs [hunk 4]
```diff
@@ -111,5 +119,63 @@ private void SetPassword(HttpClient client, string elasticUserPassword, string u
...
+			var port = httpPort ?? DefaultHttpPort;
+
+			if (string.IsNullOrEmpty(networkHost))
+				return $"http://localhost:{port}/";
+
+			var host = networkHost;
+			if (networkHost.IndexOf(',') >= 0) // Uri host must not contain comma, but elasticsearch.yml supports it
+			{
+				var hosts = networkHost.Split(new[] {','}, StringSplitOptions.RemoveEmptyEntries).Select(h => h.Trim());
+				host = hosts.FirstOrDefault(h => h == "localhost" || h == "127.0.0.1"
```
Comment (Mpdreamz) [inline src/InstallerHosts/Elastic.InstallerHosts/Elasticsearch/Tasks/Install/SetupXPackPasswordsTask.cs]: Preferring local 👍

File: src/InstallerHosts/Elastic.InstallerHosts/Elasticsearch/Tasks/Install/SetupXPackPasswordsTask.cs [hunk 5]
```diff
@@ -111,5 +119,63 @@ private void SetPassword(HttpClient client, string elasticUserPassword, string u
...
+					host = "[::1]";
+					break;
+				//TODO: differentiate between _site_ and _global_
+				case SiteAddressPlaceholder:
+				case GlobalAddressPlaceholder:
+				case SiteAddressV4Placeholder:
+				case GlobalAddressV4Placeholder:
+				case SiteAddressV6Placeholder:
+				case GlobalAddressV6Placeholder:
+					IPHostEntry ipHostEntry = Dns.GetHostEntry(Dns.GetHostName());
```
Comment (Mpdreamz) [inline src/InstallerHosts/Elastic.InstallerHosts/Elasticsearch/Tasks/Install/SetupXPackPasswordsTask.cs]: Wrap this in an try catch might be safer with a fallback to localhost?

Comment (ivan-sam) [reply src/InstallerHosts/Elastic.InstallerHosts/Elasticsearch/Tasks/Install/SetupXPackPasswordsTask.cs]: This code is to address _global_ placeholder which by definition should not use local interface. Having this in elasticsearch.yml will tell elastic server not to listen locally. So using localhost as a fallback does not make much sense to me. My idea was the following: if _global_ is used resolve of current host name should work or installation should fail with clear reason.

Comment (Mpdreamz) [reply src/InstallerHosts/Elastic.InstallerHosts/Elasticsearch/Tasks/Install/SetupXPackPasswordsTask.cs]: You are correct, the preference for local if present is handled earlier.

**Comment by Mpdreamz** [comment]
Ported to `6.7`, `7.0`, `7.x` and `master`

**Pull request merged**